### PR TITLE
Add hierarchical related ops test

### PR DESCRIPTION
**Pull request by ybc1991**

### CODE REVIEW

File: bluefog/torch/mpi_ops.py [hunk 1]
```diff
@@ -669,39 +671,47 @@ def hierarchical_neighbor_allreduce(tensor: torch.Tensor,
...
             same machine should specifiy the same weights dictionary.
-        send_neighbor_machines: The list of neighbor machines to be sent to. All processes under
-            same machine should specifiy the same machine id.
-        enable_topo_check: When send_neighbors is present, enabling this option checks if the
+        dst_machine_weights: The weights for out-neighbor machines. If set to be None, assume the
+            current node sends to all of its (out-)neighbors. If having values, assume only
+            part of (out-)neighbors will be sent to. If set to be a list, assume all the weights
+            are one. In this mode, this node sends its value to partial neighbors listed in this
+            variable in a dynamic graph, and `self_weight` and `src_weights` must be present.
+            All processes under the same machine should specifiy the same machine id.
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/torch/mpi_ops.py:689]: Done.

File: bluefog/common/mpi_context.h [hunk 2]
```diff
@@ -164,6 +164,11 @@ class MPIContext {
       const void* input_ptr, void* output_ptr, int num_elements, DataType dtype,
       const std::vector<int>* dst_ranks,
       const std::vector<int>* src_ranks);
+  std::string NeighborValueExchangeWithConstantWeightedElements(
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/common/mpi_context.h]: Done.

File: bluefog/common/mpi_context.cc [hunk 3]
```diff
@@ -490,6 +490,53 @@ std::string GenerateNeighborExchangeErrorMessage(const std::vector<MPI_Status>&
   return error_message;
 }
 
+std::string MPIContext::NeighborValueExchangeWithConstantWeightedElements(
+      const void* input_ptr, void* output_ptr, int num_elements, DataType dtype,
+      const std::vector<int>* dst_ranks, const std::vector<double>* dst_weights,
+      const std::vector<int>* src_ranks, bool dst_weighting_enabled,
+      common::ReadyEvent* ready_event, Tensor* tensor) {
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/common/mpi_context.cc]: Done.

File: bluefog/torch/mpi_ops.py [hunk 4]
```diff
@@ -715,18 +731,28 @@ def hierarchical_neighbor_allreduce_nonblocking(tensor: torch.Tensor,
     Bluefog processes for a given name. The reduction will not start until all processes
     are ready to send and receive the tensor.
 
+    In short, this API functions as the following equation.
```
Comment (ybc1991) [inline bluefog/torch/mpi_ops.py:734]: Nit: this function implements the following equation.

Comment (ybc1991) [reply bluefog/torch/mpi_ops.py:734]: Oh, functions here as verb?

File: bluefog/common/nccl_controller.cc [hunk 5]
```diff
@@ -807,16 +803,45 @@ void NCCLController::NeighborAllreduce(TensorTableEntry& entry) {
...
+                                  num_elements * i * element_size);
+          NCCLCHECK(ncclRecv(recvbuf, num_elements, GetNCCLDataType(entry.tensor),
+                             recv_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
+        }
+        for (int send_rank : *entry.send_neighbors) {
+          NCCLCHECK(ncclSend(sendbuf, num_elements, GetNCCLDataType(entry.tensor),
+                             send_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
+        }
+      } else {
+        cudaStreamSynchronize(nccl_ctx_.stream);
```
Comment (ybc1991) [inline bluefog/common/nccl_controller.cc:819]: Add a comment for the reason of adding this Sync for future reference

**Comment by ybc1991** [comment]
I cannot approve my own PR. So it should be @hanbinhu you approve it.

File: bluefog/common/nccl_controller.cc (hunk 5 above)
Comment (hanbinhu) [reply bluefog/common/nccl_controller.cc:821]: Done.

File: bluefog/torch/mpi_ops.py (hunk 4 above)
Comment (hanbinhu) [reply bluefog/torch/mpi_ops.py]: Yes. 😂  Changed anyway.

**Review by hanbinhu — APPROVED** [review]
LGTM!

**Pull request merged**